### PR TITLE
Add narrowtypes functionality

### PR DESCRIPTION
**Pull request by quinnj**
@anandijain, I'm not able to push to your branch, but I made some changes and rebased your changes here; take a look if it seems ok to you. Then we can finish your todos by adding docs/tests.

### CODE REVIEW

**Comment by anandijain** [comment]
Thanks a lot for the help! These changes all look good. 

I just added some tests, which I'll push now. I based them off of other testsets. 

Many of the other lazy structs have functions that return an anonymous function ie `x->filter(f, x)`, is this needed?

**Comment by anandijain** [comment]
Actually, I'm able to push to my fork, but I think since I don't have juliadata/ write access, I won't be able to add commits (is that right?) 

Since it's so few lines, here are the tests:
```julia 
ctable_type_any = (A=Any[1, missing, 3], B=Any[1.0, 2.0, 3.0], C=Any["hey", "there", "sailor"])

@testset "TableOperations.narrowtypes" begin

nt = TableOperations.narrowtypes(ctable_type_any)
@test Tables.istable(nt)
@test Tables.columnaccess(nt)
@test Tables.schema(nt) == Tables.schema(ctable)
@test Tables.columnnames(nt) == Tables.columnnames(ctable)

end

```

**Review by anandijain — APPROVED** [review]

**Comment by quinnj** [comment]
Thanks for the contribution @anandijain!

**Comment by anandijain** [comment]
Thanks for the help!

File: src/TableOperations.jl [hunk 1]
```diff
@@ -410,4 +410,42 @@ Base.eltype(x::RowPartitions) = Vector{MaterializedRow}
...
+narrowarray(::Type{T}, x::AbstractArray{S}) where {T, S} = Vector{T}(x)
+
+"""
+    Tables.narrowtypes(source) => Tables.NarrowTypes
+    source |> Tables.narrowtypes() => Tables.NarrowTypes
+
+Take a Tables.jl-compatible source, with potentially "wide" column types, and re-infer the schema by
+promoting the types of each actual value for each column. Useful, for example, when a columnar table source
+has a column type of `Any`, and a more concrete type is desired. Uses `Base.promote_typejoin` internally
+to do actual type promotion.
```
Comment (nalimilan) [inline src/TableOperations.jl:431]: Note that `promote_typejoin` is just `typejoin` with special handling for `Missing` and `Nothing`, so speaking about promotion is a bit misleading. Maybe using `promote` be default would be more useful? For example if you have a mix of `Int` and `Float64`, `promote_typejoin` will give `Real`, which isn't ideal.

**Pull request merged**